esp-hal/src/mcpwm/mod.rs

```diff
@@ -101,6 +102,29 @@ pub mod timer;
 
 type RegisterBlock = pac::mcpwm0::RegisterBlock;
 
+#[allow(dead_code)] // Field is seemingly unused but we rely on its Drop impl
+struct PwmClockGuard(OnDrop<fn()>);
+
+impl PwmClockGuard {
+    pub fn new<PWM: PwmPeripheral>() -> Self {
+        if PWM::peripheral() == Peripheral::Mcpwm0 {
+            ClockTree::with(clocks::request_mcpwm0_function_clock);
+        } else {
+            #[cfg(soc_has_mcpwm1)]
+            ClockTree::with(clocks::request_mcpwm1_function_clock);
+        }
+
+        Self(OnDrop::new(|| {
+            if PWM::peripheral() == Peripheral::Mcpwm0 {
+                ClockTree::with(clocks::release_mcpwm0_function_clock);
+            } else {
+                #[cfg(soc_has_mcpwm1)]
+                ClockTree::with(clocks::release_mcpwm1_function_clock);
+            }
+        }))
+    }
+}
```

Comment:
I admit this is nasty enough to replace with something else. I'm considering separate clock node types, but I'm unsure if we should just use the config types or create new ones.